### PR TITLE
chore: introduce ci

### PR DESCRIPTION
**Pull request by Orkuncakilkaya**

### CODE REVIEW

**Review by TheUnderScorer — COMMENTED** [review]
Question: wouldn't it be easier to deploy the service using [Terraform](https://registry.terraform.io/providers/fastly/fastly/latest/docs)?

File: .github/workflows/mock-e2e-pr.yml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+      - name: Install dependencies
+        run: pnpm install
+      - name: Extract Branch Name
+        id: extract-branch
+        run: |
+          echo SUBDOMAIN=mock-$(echo ${GITHUB_HEAD_REF:-${GITHUB_REF#refs/heads/}} | perl -pe 's/[^a-zA-Z0-9]+/-/g and s/-+$//g' | tr '[:upper:]' '[:lower:]') >> $GITHUB_OUTPUT
+      - name: Deploy
+        id: deploy
+        run: |
+          SERVICE_NAME="${{steps.extract-branch.outputs.SUBDOMAIN}}" DEFAULT_ORIGIN=${{secrets.DEFAULT_ORIGIN}} FASTLY_API_TOKEN="${{secrets.FASTLY_API_TOKEN}}" FPJS_BACKEND_URL="${{secrets.MOCK_FPCDN}}" FPCDN_BACKEND_URL="${{secrets.MOCK_FPCDN}}" CONFIG_STORE_NAME="${{steps.extract-branch.outputs.SUBDOMAIN}}" pnpm run ci
```
[the remark this answers is not in the document]
Comment (Orkuncakilkaya) [reply .github/workflows/mock-e2e-pr.yml]: Done 👍

File: scripts/api/createService.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+export async function createService(domain: string) {
+  const client = createClient('service')
+  try {
+    const searchResponse = await client.searchService({ name: domain })
+    if (searchResponse && searchResponse.id) {
+      return client.getServiceDetail({ service_id: searchResponse.id })
+    }
+  } catch (_) {
+    console.log(`Couldn't find service with name: ${domain}`)
```
[the remark this answers is not in the document]
Comment (Orkuncakilkaya) [reply scripts/api/createService.ts]: Done 👍

**Comment by Orkuncakilkaya** [comment]
> Question: wouldn't it be easier to deploy the service using [Terraform](https://registry.terraform.io/providers/fastly/fastly/latest/docs)?

It would be easier to maintain, but would be hard to develop at first. Some parts were already developed for VCL. Starting from scratch via terraform would take longer

File: .github/workflows/mock-e2e-pr.yml [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+jobs:
+  check-changed-files:
+    runs-on: ubuntu-20.04
+    name: Check changed files
+    outputs:
+      CANT_PASS_WITHOUT_TESTS: ${{ steps.decision.outputs.CANT_PASS_WITHOUT_TESTS }}
+    steps:
+      - id: files
+        uses: jitterbit/get-changed-files@b17fbb00bdc0c0f63fcf166580804b4d2cdc2a42
```
Comment (necipallef) [inline .github/workflows/mock-e2e-pr.yml:14]: This is coming from Akamai. Let's use the one from CloudFront because this library is not receiving any updates for a long time.

File: scripts/api/activateVersion.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import { createClient } from '../utils/createClient'
+
+export async function activateVersion(service_id: string, versionId: number) {
```
Comment (necipallef) [inline scripts/api/activateVersion.ts:3]: nitpick: no need for `async` here

File: scripts/api/createService.ts [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    service_id,
+    version_id,
+    address: process.env.FPJS_BACKEND_URL ?? 'api.fpjs.io',
+    name: 'fpjs',
+    port: 443,
+  })
+  await client.createBackend({
+    service_id,
+    version_id,
+    address: process.env.FPCDN_BACKEND_URL ?? 'fpcdn.io',
```
Comment (necipallef) [inline scripts/api/createService.ts:129]: nitpick: I'd go with the name `FPCDN_BACKEND_URL` instead of `FPCDN_URL` because it is not a traditional backend, but this is not _that_ important.

File: build.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
+const esbuild = require('esbuild')
```
Comment (necipallef) [inline build.js:1]: Could you briefly explain why we use `esbuild` instead of `rollup` like we do in other projects?

File: tsconfig.json [hunk 7]
```diff
@@ -14,9 +14,16 @@
...
     "./test/**/*.ts"
   ],
   "exclude": [
     "node_modules"
-  ]
+  ],
+  "ts-node": {
+    "experimentalSpecifierResolution": "node",
+    "compilerOptions": {
+      "module": "CommonJS"
```
Comment (necipallef) [inline tsconfig.json:26]: Could you briefly explain why this change was needed? The whole `ts-node` object

**Review by necipallef — CHANGES_REQUESTED** [review]

File: tsconfig.json (hunk 7 above)
Comment (Orkuncakilkaya) [reply tsconfig.json:26]: I would like to have different options than tsconfig.json when I want to run my scripts with ts-node. But I wouldn't like to change all tsconfig for that, so instead I used ts-node option only available for running ts-node environment.

File: build.js (hunk 6 above)
Comment (Orkuncakilkaya) [reply build.js:1]: Lot's of reasons here I used esbuild.

1. This project immediately compiles built code from JavaScript to WebAssembly, so there is no need for plugins like banner, and therefore no need to directly use our rollup config.

2. We already use 2 steps of builds, and both have their own compile time, but esbuild comes to rescue because it's far faster than rollup and other bundlers. See this image for how fast they are:
image.
You can get detailed information from their website and repository.

3. Rollup needs external plugins for commonjs transpiler, typescript loader and etc. With esbuild these kind of plugins comes in the box. As you can see in this PR I didn't add any loader or transpiler plugins for esbuild and speed up my development process.

File: scripts/api/createService.ts (hunk 5 above)
Comment (Orkuncakilkaya) [reply scripts/api/createService.ts]: This commit refactors it: https://github.com/fingerprintjs/fingerprint-pro-fastly-proxy-integration-example/pull/7/commits/5ec6135dc3a6a23be6413f407fdefad2dde93aac

File: .github/workflows/mock-e2e-pr.yml (hunk 3 above)
Comment (Orkuncakilkaya) [reply .github/workflows/mock-e2e-pr.yml]: This commit fixes it: https://github.com/fingerprintjs/fingerprint-pro-fastly-proxy-integration-example/pull/7/commits/6758dcc7654b2e0fcbc9b46e4bf5b33b86c1e728

File: scripts/api/activateVersion.ts (hunk 4 above)
Comment (Orkuncakilkaya) [reply scripts/api/activateVersion.ts]: This commit refactors it: https://github.com/fingerprintjs/fingerprint-pro-fastly-proxy-integration-example/pull/7/commits/5ec6135dc3a6a23be6413f407fdefad2dde93aac

File: build.js (hunk 6 above)
Comment (ilfa) [reply build.js:1]: 1. Do we plan to distribute only wasm file? I also noticed that we generate `build/index.js`
2. Are these tests for random project or our? I mean if it is less than a few seconds and we are not going to grow the project significantly it is not a correct criteria.
3. If an instrument needs plugins to do something it could also be a good thing because in this case esbuild looks like a Swiss knife and it means that we have more hidden deps.

I do not offer you to change the instrument, just noticed weak arguments for exact case.

Comment (Orkuncakilkaya) [reply build.js:1]: 1. There is no need to distribute the build/index.js file because it is of no use; it's just built because js-compute-runtime requires javascript instead of typescript to build. 
2. These benchmark tests are not for our repository but for a medium-weight repository like us.
3. But "esbuild" handles those plugins' maintenance by itself. For example, when we switch to a new version of typescript, we won't need to update plugins and check their compatibility with each other.

Comment (ilfa) [reply build.js:1]: 3. so we use a package with an uncontrolled amount of dependencies that could change at any time? =)

**Comment by ilfa** [comment]
@Orkuncakilkaya please check your local version of pnpm?  I see that that lock file has version 6 (actual is 9)

File: build.js (hunk 6 above)
Comment (Orkuncakilkaya) [reply build.js:1]: 3. I think it's not that simple and I think this package has so much users, should strongly depend on semver

**Review by necipallef — APPROVED** [review]

**Review by TheUnderScorer — APPROVED** [review]

File: build.js (hunk 6 above)
Comment (necipallef) [reply build.js:1]: Instead of using `esbuild` or `rollup`, I suggest a 3rd option: do not use anything at all.

We can rely on running `tsc` to transpile to JS, and `fastly compute build` to generate wasm. [Here](https://github.com/fastly/compute-starter-kit-typescript/blob/main/package.json) is an example Typescript project of Fastly.

The only thing we would lose is the ability to change config store name, right? But on the other hand, we wouldn't need any configuration file, `build.ts` or `rollup.config.js`, no dependency on any build system, we can even remove the entries for webpack from `package.json`. I think this is worth it. And we can use the hardcoded config store name because currently we are not providing the customer to change this name anyway.

Pls let me know what you think @ilfa @Orkuncakilkaya

Comment (Orkuncakilkaya) [reply build.js:1]: We need to use different Config Store Names for PR E2E Tests and Daily E2E Tests. So we require a build tool or script that will provide us this feature. And the easiest I could find was esbuild. We can switch to rollup for consistency or we can find another way with tsc or something else but IMHO it's really not worth the effort for the case.

Comment (TheUnderScorer) [reply build.js]: #1 is okay for me as well

Comment (ilfa) [reply build.js]: #1 is ok for me in this project.

File: src/index.ts [hunk 8]
```diff
@@ -4,10 +4,17 @@ import { IntegrationEnv } from './env'
...
 import { createFallbackErrorResponse } from './utils'
+import { setClientIp } from './utils/clientIp'
 
 addEventListener('fetch', (event) => event.respondWith(handleRequest(event)))
 
 export async function handleRequest(event: FetchEvent): Promise<Response> {
+  setClientIp(
+    event.client.address ??
+      event.request.headers.get('Fastly-Client-IP') ??
+      event.request.headers.get('X-Forwarded-For') ??
```
Comment (necipallef) [inline src/index.ts:15]: I think we should remove `XFF` header because it is spoofable. `Fastly-Client-IP` looks to be true client IP according to [here](https://www.fastly.com/documentation/reference/http/http-headers/Fastly-Client-IP/) so that seems fine but I'd also like to see if it is possible spoof this header as well. Could you explain why you decided to use this header instead of [this usage](https://github.com/fingerprintjs/fingerprint-pro-fastly-proxy-integration-example/blob/main/src/index.ts#L13) we had before? I don't remember how we came up with this implementation before but would like to know why you picked the header over this @Orkuncakilkaya

File: build.js (hunk 6 above)
Comment (JuroUhlar) [reply build.js]: Yes, number 1 is good!

Comment (necipallef) [reply build.js]: Thanks for the feedback! @Orkuncakilkaya feel free to resolve the conversation

File: src/index.ts (hunk 8 above)
Comment (Orkuncakilkaya) [reply src/index.ts:15]: I applied this solution as a failover if one of them fails, next one should work. In order, our old way works and probably second condition will never work. But if a glitch happens we will have a failover scenario that won't affect identification. I can simply remove those and apply the old way but I think it's best to leave as it is because without a glitch in the system, it'll never work.

Comment (necipallef) [reply src/index.ts:15]: In that case it is better to let it fail than picking an un-safe value. I'd remove any failovers

Comment (Orkuncakilkaya) [reply src/index.ts]: Done 👍

**Review by necipallef — APPROVED** [review]
